linux/mcompile/keymap/keymap.cpp

```diff
@@ -17,10 +17,10 @@ int convert_Shiftstate_to_LinuxShiftstate(int shiftState) {
 
 
 bool ensureValidInputForKeyboardTranslation(int gdk_level, gint count, gint keycode) {
-  if (gdk_level >=  (int) count)
+  if (gdk_level >  (int) count)
```

Comment:
We're going back and forth with this. This should still be `>=` for reasons discussed previously.

```suggestion
  if (gdk_level >= (int) count)
```

Reply:
`(gdk_level >  (int) count) is correct.` Otherwise we will miss some values for SPACE:


![image](https://github.com/keymanapp/keyman/assets/86713187/7146faf1-0895-4687-8ab5-93cc28c21a58)

 (gdk_level >= count)  would be correct for all Keys  -**EXCEPT** for SPACE: For SPACE we would not find the values for shift:

- Linux-Shiftstate Shift:  `Gdk_level = 1`
- for K_SPACE gdk_keymap_get_entries_for_keycode calculates `count  = 1`
- `if (gdk_level >= count)   returns false` in this case => `Shift+SPACE is not processed`

[From GDK:  some keys don’t have multiple groups - e.g. the Enter key is always in group 0 regardless of keyboard state.](https://docs.gtk.org/gdk4/keys.html)

Reply:
Then we're probably comparing apples and oranges. `gdk_keymap_get_entries_for_keycode` allocates memory for `maps` and `keyvals` and returns the size in `count`. For `maps` and `keyvals` the condition has to be `n < count`.

I don't understand the relationship between `gdk_level` and `count`.

Reply:
```
On a Linux Keyboard we can have up to 4 Groups e.g.German, French, English, Spanish.
Every Key in a Group can have up to 4 levels ( count =4):
   level=0 - BASE
   level=1 - SHIFT
   level=2 - ALT
   level=3 - SHIFT+ALT

But for key K_SHIFT we have count=1 and use level 0 and level 1
```

- For normal keys we can compare all Shiftstates (0-3) since n=0<16;    n=1<16;    n=2<16;   n=3<16

- Even if we only had one language it would still be  n=0<4;    n=1<4;     n=2<4;   n=3<4

- **BUT for K_Space this will not work** since we want Base and Shift and always have count=1  :   n=0<1;   **n=1<1;** 

Therefore because of K_Space we need to use n <= count     and to exclude:  ` if (gdk_level >  (int) count)`
___________________________________________
```
K_Space:
n=0, keycode=65, keyval=32 ( ), level=0, group=0
```


```
German  K_Z
    n=0, keycode=52, keyval=121 (y), level=0, group=0
    n=1, keycode=52, keyval=89 (Y), level=1, group=0
    n=2, keycode=52, keyval=187 (»), level=2, group=0
    n=3, keycode=52, keyval=8250 (›), level=3, group=0
```
```
French(AZERTY)  K_Z
    n=4, keycode=52, keyval=119 (w), level=0, group=1
    n=5, keycode=52, keyval=87 (W), level=1, group=1
    n=6, keycode=52, keyval=0 (), level=2, group=1
    n=7, keycode=52, keyval=0 (), level=3, group=1
```

```
English(US)  K_Z
    n=8, keycode=52, keyval=122 (z), level=0, group=2
    n=9, keycode=52, keyval=90 (Z), level=1, group=2
    n=10, keycode=52, keyval=0 (), level=2, group=2
    n=11, keycode=52, keyval=0 (), level=3, group=2
```

```
Spanish  K_Z
    n=12, keycode=52, keyval=122 (z), level=0, group=3
    n=13, keycode=52, keyval=90 (Z), level=1, group=3
    n=14, keycode=52, keyval=0 (), level=2, group=3
    n=15, keycode=52, keyval=0 (), level=3, group=3
```

Reply:
I'm beginning to understand how this is supposed to work and why it smells. We basically test that `gdk_level` is less or equal than the maximum possible level (that we're interested in), except for a Space where `gdk_level` can only be 0 or 1. Comparing `gdk_level >= count` is a (dirty) shortcut which happens to work if we assume that `gdk_level` is in the range `0..3`. It would work with other values as well because we don't do anything with `count` elsewhere, but conceptually it's wrong.

I think we have to make this a bit more explicit so that it's easier to understand.

Do we know the keycode for `K_Space`? Then we could do something like

```
// We're only dealing with levels 0-3
if (gdk_level < 0 || gdk_level > 3)
  return false;
// For K_Space only non-Shift and Shift are allowed
if (keycode == K_Space && gdk_level > 1)
  return false;
```

If we don't know the keycode for `K_Space` then we could do:

```
// We're only dealing with levels 0-3
if (gdk_level < 0 || gdk_level > 3)
  return false;
// K_Space has only one level in the keyboard, but we want to allow non-Shift and Shift
if (gdk_level > 1 && count <= 3)
  return false;
```

Reply:
OK yet another change: 
In the first comparison I changed `gdk_level >3` to `shiftstate > 7` and renamed the variable since ensureValidInputForKeyboardTranslation() is used  with shiftstates
  - 0,1,2,3 for Linux-SS and
  - 0,1,6,7 for rgkey-SS 
 
both meaning Base, Shift, AltGr, Sh+AltGr